0006-bingings.md

```diff
@@ -0,0 +1,584 @@
+```
+All contract calls are a result of a receipt, this receipt might be created by a transaction
+that does function invocation on the contract or another contract as a result of cross-contract call.
+
+###### Normal operation
+* Saves the bytes of the predecessor account id into the register.
+
+###### Panics
+* If the registers exceed the memory limit panics with `MemoryAccessViolation`;
+
+###### Current bugs
+* Not implemented.
+
+---
+```rust
+input(register_id: u64)
+```
+Reads input to the contract call into the register. Input is expected to be in JSON-format.
+
+###### Normal operation
+* If input is provided saves the bytes (potentially zero) of input into register.
+* If input is not provided does not modify the register.
+
+###### Returns
+* If input was not provided returns `0`;
+* If input was provided returns `1`; If input is zero bytes returns `1`, too.
+
+###### Panics
+* If the registers exceed the memory limit panics with `MemoryAccessViolation`;
+
+###### Current bugs
+* Implemented as part of `data_read`. However there is no reason to have one unified function, like `data_read` that can
+be used to read all
+
+---
+```rust
+block_index() -> u64
+```
+Returns the current block index.
+
+---
+```rust
+storage_usage() -> u64
+``` 
+Returns the number of bytes used by the contract if it was saved to the trie as of the
+invocation. This includes:
+* The data written with `storage_*` functions during current and previous execution;
+* The bytes needed to store the account protobuf and the access keys of the given account.
+
+## Economics API
+Accounts own certain balance; and each transaction and each receipt have certain amount of balance and prepaid gas
+attached to them.
+During the contract execution, the contract has access to the following `u128` values:
+* `account_balance` -- the balance attached to the given account. This includes the `attached_deposit` that was attached
+  to the transaction;
+* `attached_deposit` -- the balance that was attached to the call that will be immediately deposited before
+  the contract execution starts;
+* `prepaid_gas` -- the tokens attached to the call that can be used to pay for the gas;
+* `used_gas` -- the gas that was already burnt during the contract execution and attached to promises (cannot exceed `prepaid_gas`);
+
+If contract execution fails `prepaid_gas - used_gas` is refunded back to `signer_account_id` and `attached_balance`
+is refunded back to `predecessor_account_id`.
+
+The following spec is the same for all functions:
+```rust
+account_balance(balance_ptr: u64)
+attached_deposit(balance_ptr: u64)
+
+```
+ -- writes the value into the `u128` variable pointed by `balance_ptr`.
+
+###### Panics
+* If `balance_ptr + 16` points outside the memory of the guest with `MemoryAccessViolation`;
+
+###### Current bugs
+* Use a different name;
+
+---
+```rust
+prepaid_gas() -> u64
+used_gas() -> u64
+```
+
+## Math
+
+```rust
+random_seed(register_id: u64)
```

Comment:
Why does this need to take register_id?

Reply:
That's how we pass blobs between host and guest.

Reply:
Do we need this for every function? It is somewhat confusing from outside. Maybe we can agree for certain functions which registers we write to.

Reply:
Blockchain Wasm APIs are in general ugly, we make it pretty by wrapping it with `near.ts`. Motivation on why we need registers is provided in the registers sections of this doc. It applies to every function that operates with blobs.

Also, this API is not subject to modifications anymore. We had several last call announcements for changes two weeks ago.